spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/PushDown.scala

```diff
@@ -0,0 +1,204 @@
+package org.apache.spark.sql.cassandra
+
+import com.datastax.spark.connector.cql.TableDef
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.sources
+import org.apache.spark.sql.sources.Filter
+
+/**
+ * Calculate pushdown predicates for a given table and Catalyst [[Expression]]s
+ */
+class PredicatePushDown (predicates: Seq[Expression], tableDef: TableDef)
+  extends PushDown(predicates: Seq[Expression], tableDef: TableDef) {
+
+  def isEqualTo(predicate: Expression): Boolean = predicate.isInstanceOf[EqualTo]
+
+  def isIn(predicate: Expression) : Boolean = predicate.isInstanceOf[In] || predicate.isInstanceOf[InSet]
+
+  def isRangeComparison(predicate: Expression) : Boolean = predicate match {
+    case _: LessThan           => true
+    case _: LessThanOrEqual    => true
+    case _: GreaterThan        => true
+    case _: GreaterThanOrEqual => true
+    case _                     => false
+  }
+
+  override def isSingleColumn(predicate: Expression) : Boolean  =
+    predicate.references.size == 1 && super.isSingleColumn(predicate)
+
+  /** Returns the only column name referenced in the predicate */
+  def predicateColumnName(predicate: Expression) : String = {
+    require(predicate.references.size == 1, "Given predicate is not a single column predicate: " + predicate)
+    predicate.references.head.name
+  }
+}
+
+/**
+ * Calculate pushdown filters for a given table and source [[Filter]]s
+ */
+class FilterPushDown (filters: Seq[Filter], tableDef: TableDef)
+  extends PushDown(filters: Seq[Filter], tableDef: TableDef) {
+
+  def isEqualTo(filter: Filter): Boolean = filter.isInstanceOf[sources.EqualTo]
+
+  def isIn(filter: Filter) : Boolean = filter.isInstanceOf[sources.In]
+
+  def isRangeComparison(filter: Filter) : Boolean = filter match {
+    case _: sources.LessThan           => true
+    case _: sources.LessThanOrEqual    => true
+    case _: sources.GreaterThan        => true
+    case _: sources.GreaterThanOrEqual => true
+    case _                             => false
+  }
+
+  /** Returns the only column name referenced in the predicate */
+  def predicateColumnName(filter: Filter) : String = {
+    filter match {
+      case eq: sources.EqualTo            => eq.attribute
+      case lt: sources.LessThan           => lt.attribute
+      case le: sources.LessThanOrEqual    => le.attribute
+      case gt: sources.GreaterThan        => gt.attribute
+      case ge: sources.GreaterThanOrEqual => ge.attribute
+      case in: sources.In                 => in.attribute
+      case _ =>
+        throw new UnsupportedOperationException(
+          "It's not a valid filter to be pushed down, only >, <, >=, <= and In are allowed: " + filter)
```

Comment:
Same idea here:

s"'filter' $filter is not a valid pushdown format, must be >, <, >=, <= and In are allowed")

or something like that IMO